src/goto-programs/remove_java_nondet.cpp

```diff
@@ -0,0 +1,389 @@
+
+ Purpose: Return whether the instruction is an assignment, and the rhs is a
+          symbol or typecast expression with the specified identifier.
+
+\*******************************************************************/
+
+static bool is_assignment_from(
+  const goto_programt::instructiont &instr,
+  const dstringt &identifier)
+{
+  // If not an assignment, return false
+  if(!instr.is_assign())
+  {
+    return false;
+  }
+  const auto &rhs=to_code_assign(instr.code).rhs();
+  return is_symbol_with_id(rhs, identifier) ||
+         is_typecast_with_id(rhs, identifier);
+}
+
+/*******************************************************************\
+
+Function: process_target
+
+  Inputs:
+    message_handler: Handles logging.
+    symbol_table: The table of known symbols.
+    max_nondet_array_length: Max length of arrays in new nondet objects.
+    instructions: The list of all instructions.
+    instr: The instruction to check.
+
+ Outputs: The next instruction to process, probably with this function.
+
+ Purpose: Given an iterator into a list of instructions, modify the list to
+          replace 'nondet' library functions with CBMC-native nondet
+          expressions, and return an iterator to the next instruction to check.
+
+\*******************************************************************/
+
+static goto_programt::targett process_target(
+  message_handlert &message_handler,
+  symbol_tablet &symbol_table,
+  const size_t max_nondet_array_length,
+  goto_programt::instructionst &instructions,
+  const goto_programt::targett &instr)
+{
+  // Check whether this is our nondet library method, and return if not
+  const auto instr_info=get_nondet_instruction_info(instr);
+  if(!instr_info.is_nondet_instruction)
+  {
+    return std::next(instr);
+  }
+
+  // Look at the next instruction, ensure that it is an assignment
+  const auto next_instr=std::next(instr, 1);
+  assert(next_instr->is_assign());
+  // Get the name of the LHS of the assignment
+  const auto return_identifier=
+    to_code_assign(next_instr->code).lhs().get(ID_identifier);
+
+  const auto end=std::end(instructions);
+
+  // Look for an instruction where this name is on the RHS of an assignment
+  const auto matching_assignment=std::find_if(
+      next_instr,
+      end,
+      [&return_identifier](const goto_programt::instructiont &instr)
+      {
+        return is_assignment_from(instr, return_identifier);
+      });
+
+  assert(matching_assignment!=end);
+
+  // Assume that the LHS of *this* assignment is the actual nondet variable
+  const auto &code_assign=to_code_assign(matching_assignment->code);
+  const auto nondet_var=code_assign.lhs();
+  const auto source_loc=instr->source_location;
+
+  // Erase from the nondet function call to the assignment
+  const auto after_matching_assignment=std::next(matching_assignment);
+  assert(after_matching_assignment!=end);
+  const auto after_erased=
+    instructions.erase(instr, after_matching_assignment);
+
```

Comment:
Hacking the internals of a goto_program directly is not a good idea; use the methods it provides.